### PR TITLE
follow redirects in gatling test  manually when using oauth

### LINKED ISSUE #8552: Gatling with OAuth2 and Keycloak
##### **Overview of the issue**

Gatling tests don't work with OAuth2 and Keycloak

##### **Motivation for or Use Case**

Here the stacktrace:

```
================================================================================
2018-10-13 10:06:39                                          90s elapsed
---- Requests ------------------------------------------------------------------
> Global                                                   (OK=19397  KO=0     )
> First unauthenticated request                            (OK=100    KO=0     )
> Authentication                                           (OK=100    KO=0     )
> Authentication Redirect 1                                (OK=100    KO=0     )
> patternfly.css Redirect 1                                (OK=100    KO=0     )
> favicon.ico Redirect 1                                   (OK=100    KO=0     )
> patternfly-additions.css Redirect 1                      (OK=100    KO=0     )
> zocial.css Redirect 1                                    (OK=100    KO=0     )
> login.css Redirect 1                                     (OK=100    KO=0     )
> Authenticate                                             (OK=100    KO=0     )
> Authenticate Redirect 1                                  (OK=100    KO=0     )
> Authenticate Redirect 2                                  (OK=100    KO=0     )
> Authenticate Redirect 3                                  (OK=100    KO=0     )
> Authenticate Redirect 4                                  (OK=100    KO=0     )
> Authenticate Redirect 5                                  (OK=100    KO=0     )
> Authenticate Redirect 6                                  (OK=100    KO=0     )
> Authenticate Redirect 7                                  (OK=100    KO=0     )
> Authenticate Redirect 8                                  (OK=100    KO=0     )
> Authenticate Redirect 9                                  (OK=100    KO=0     )
> Authenticate Redirect 10                                 (OK=100    KO=0     )
> Authenticate Redirect 11                                 (OK=100    KO=0     )
> Authenticate Redirect 12                                 (OK=100    KO=0     )
> Authenticate Redirect 13                                 (OK=100    KO=0     )
> Authenticate Redirect 14                                 (OK=100    KO=0     )
> Authenticate Redirect 15                                 (OK=100    KO=0     )
```

##### **Reproduce the error**

I generated the project with JHipster Online:

```
git clone https://github.com/pascalgrimaud/jhipster-test-oauth2
cd jhipster-test-oauth2
npm install
docker-compose -f src/main/docker/keycloak.yml up -d
./mvnw
```

In another Terminal:

```
gatling -sf src/test/gatling/user-files/simulations
```

##### **JHipster Version(s)**

5.4.2 and current master

##### **JHipster Version(s)**

```
jhipster@0.0.0 /home/pgrimaud/tmp/jhipster-test-oauth2
└── generator-jhipster@5.4.2 

```

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**

<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "applicationType": "monolith",
    "gitCompany": "",
    "baseName": "jhipster",
    "packageName": "io.github.jhipster.application",
    "packageFolder": "io/github/jhipster/application",
    "serverPort": 8080,
    "serviceDiscoveryType": false,
    "authenticationType": "oauth2",
    "uaaBaseName": "../uaa",
    "cacheProvider": "ehcache",
    "enableHibernateCache": true,
    "websocket": false,
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "prodDatabaseType": "mysql",
    "searchEngine": false,
    "enableSwaggerCodegen": false,
    "messageBroker": false,
    "buildTool": "maven",
    "useSass": false,
    "clientPackageManager": "npm",
    "testFrameworks": [
      "gatling",
      "protractor"
    ],
    "enableTranslation": false,
    "nativeLanguage": "en",
    "languages": [
      "en"
    ],
    "clientFramework": "angularX",
    "jhiPrefix": "jhi",
    "jhipsterVersion": "5.4.2"
  },
  "git-provider": "GitHub",
  "git-company": "pascalgrimaud",
  "repository-name": "jhipster-test-oauth2"
}
</pre>
</details>

##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

<details>
<summary>JDL entity definitions</summary>

<pre>
entity Toto {
  name String
}

</pre>
</details>

##### **Environment and Tools**

openjdk version "1.8.0_181"
OpenJDK Runtime Environment (build 1.8.0_181-8u181-b13-0ubuntu0.18.04.1-b13)
OpenJDK 64-Bit Server VM (build 25.181-b13, mixed mode)

git version 2.19.1

node: v8.12.0

npm: 6.4.1

yeoman: 2.0.5

yarn: 1.10.1

Docker version 18.03.0-ce, build 0520e24

docker-compose version 1.21.2, build a133471
-   [x] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by atomfrede**
closes #8552

-   Please make sure the below checklist is followed for Pull Requests.

-   [ ] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
-   [x] Tests are added where necessary
-   [x] Documentation is added/updated where necessary
-   [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

File: generators/entity-server/templates/src/test/gatling/user-files/simulations/EntityGatlingTest.scala.ejs [hunk 1]
```diff
@@ -47,7 +47,9 @@ class <%= entityClass %>GatlingTest extends Simulation {
         .acceptEncodingHeader("gzip, deflate")
         .acceptLanguageHeader("fr,fr-fr;q=0.8,en-us;q=0.5,en;q=0.3")
         .connectionHeader("keep-alive")
-        .userAgentHeader("Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:33.0) Gecko/20100101 Firefox/33.0")
+        .userAgentHeader("Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:33.0) Gecko/20100101 Firefox/33.0")<%_ if (authenticationType === 'oauth2') { _%>
```
Comment (pascalgrimaud) [inline generators/entity-server/templates/src/test/gatling/user-files/simulations/EntityGatlingTest.scala.ejs]: the condition with `<%_` should be in a new line

**Comment by pascalgrimaud** [comment]
I will test it very soon. 
Hope it will work. Nice work @atomfrede

**Comment by pascalgrimaud** [comment]
I tested with Monolith -> it works fine
I tested with Microservices + Consul -> I got some errors (302 instead of 401) but I know how to fix it. Give me some time, I should do suggestion to this PR tonight or tomorrow

**Comment by pascalgrimaud** [comment]
PR done: https://github.com/atomfrede/generator-jhipster/pull/1 with some explainations

For your information, I tested 5 use cases:
- monolithic with OAuth2 + Keycloak
    - start normally
    - start with docker-compose
- gateway with OAuth2 + Keycloak
    - start normally
    - start with docker-compose
- microservice with gateway + oauth2 + keycloak + consul
    - start the full stack with docker-compose

Everything works fine for me. Thanks a lot @atomfrede !

**Comment by atomfrede** [comment]
:clap: That is teamwork :+1:

**Comment by pascalgrimaud** [comment]
merged ! don't forget to claim the bounty, @atomfrede, this one is well deserved !

**Comment by atomfrede** [comment]
Bounty claimed https://opencollective.com/generator-jhipster/expenses/5318

**Pull request merged**